File: lumea-auth/internal/middleware/apikey.go
package middleware

import (
	"net/http"

	"github.com/gin-gonic/gin"
)

// RequireAPIKey validates the X-API-Key header against the provided set of valid keys.
// Keys come from config.APIKeys (split from the API_KEYS env var on "-").
// /health and /docs/* are always exempted so infra health checks still work.
func RequireAPIKey(validKeys []string) gin.HandlerFunc {
	keySet := make(map[string]struct{}, len(validKeys))
	for _, k := range validKeys {
		if k != "" {
			keySet[k] = struct{}{}
		}
	}

	return func(c *gin.Context) {
		path := c.Request.URL.Path

		// Exempt health check and Swagger — these don't carry a BFF key
		if path == "/health" || len(path) >= 5 && path[:5] == "/docs" {
			c.Next()
			return
		}

		// If no keys are configured (e.g. local dev without env set), skip check
		if len(keySet) == 0 {
			c.Next()
			return
		}

		key := c.GetHeader("X-API-Key")
		if _, ok := keySet[key]; !ok {
			c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "Invalid or missing API key"})
			return
		}

		c.Next()
	}
}

File: lumea-analytics/src/middleware/apikey.ts
import { createMiddleware } from "hono/factory";

const EXEMPT = new Set(["/health", "/docs", "/openapi.json"]);

export function requireAPIKey() {
  return createMiddleware(async (c, next) => {
    if (EXEMPT.has(c.req.path)) return next();

    const raw = process.env.API_KEYS ?? "";
    if (!raw) return next(); // no keys configured — open in dev

    const validKeys = new Set(raw.split("-").filter(Boolean));
    const provided = c.req.header("X-API-Key") ?? "";

    if (!validKeys.has(provided)) {
      return c.json({ error: "Invalid or missing API key" }, 401);
    }

    return next();
  });
}
